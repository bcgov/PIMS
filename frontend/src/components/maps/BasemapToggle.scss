@import '../../colors.scss';

.basemap-container {
  position: absolute;
  bottom: 0;
  left: 0;
  border-radius: 8px;
  border: 2px solid white;
  box-shadow: 0px 1px 4px rgba(0, 0, 0, 0.3);
  background-color: #ccc;
  color: $light-variant-color;
  width: 75px;
  height: 75px;
  margin: 1em;
  transition: opacity 0.25s ease-in-out;
  z-index: 600;

  .caption {
    pointer-events: none;
    position: absolute;
    bottom: 0;
    left: 0;
    width: 100%;
    border-radius: 6px;
    cursor: pointer;
    color: white;
    font-size: 12px;
    line-height: normal;
    padding: 12px 0px 6px 0px;
    margin-bottom: 0;
    background-image: linear-gradient(transparent, rgba(0, 0, 0, 0.6));
    text-indent: 6px;
    text-shadow: rgba(0, 0, 0, 0.7) 0px 1px 8px;
    text-align: left;
    z-index: 590;
  }

  .basemap-item {
    position: absolute;
    bottom: 0;
    left: 0;
    top: 0;
    right: 0;
    border-radius: 8px;
    overflow: hidden;
    z-index: 550;
    background-color: rgba(255, 255, 255, 0.01);
    display: none;
    &.secondary {
      z-index: 551;
      display: block;
    }
    &-button {
      cursor: pointer;
    }
  }
}
