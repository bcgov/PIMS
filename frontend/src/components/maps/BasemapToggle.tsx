import './BasemapToggle.scss';

import React, { useState } from 'react';

export type BaseLayer = {
  name: string;
  url: string;
  attribution: string;
  thumbnail: string;
};

export type BasemapToggleEvent = {
  current: BaseLayer;
  previous: BaseLayer;
};

type BasemapToggleProps = {
  baseLayers: BaseLayer[];
  onToggle?: (e: BasemapToggleEvent) => void;
};

const BasemapToggle: React.FC<BasemapToggleProps> = props => {
  // TODO: fade out when map is panning/zooming
  const [updating] = useState(false);
  const [currentBasemap, secondaryBasemap] = props.baseLayers;

  const toggle = () => {
    const e: BasemapToggleEvent = {
      current: secondaryBasemap,
      previous: currentBasemap,
    };
    props.onToggle?.(e);
  };

  return (
    <div className={updating ? 'basemap-container view-busy' : 'basemap-container'}>
      <button className="basemap-item current"></button>
      <label className="caption">{secondaryBasemap?.name}</label>
      <div className="basemap-item basemap-item-button secondary" onClick={toggle}>
        <img src={secondaryBasemap?.thumbnail} role="presentation" alt="Map Thumbnail" />
      </div>
    </div>
  );
};

export default BasemapToggle;
