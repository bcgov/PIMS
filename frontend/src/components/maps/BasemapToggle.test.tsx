import Adapter from '@cfaester/enzyme-adapter-react-18';
import { mount } from 'enzyme';
import Enzyme from 'enzyme';
import React from 'react';
import renderer from 'react-test-renderer';

import BasemapToggle, { BaseLayer } from './BasemapToggle';

Enzyme.configure({ adapter: new Adapter() });

const toggle = jest.fn();

const baseMaps = [
  {
    name: 'Map',
    url: 'https://{s}.tile.openstreetmap.org/{z}/{x}/{y}.png',
    attribution: '&copy; <a href="http://osm.org/copyright">OpenStreetMap</a> contributors',
    thumbnail: 'streets.jpg',
  },
  {
    name: 'Satellite',
    url: 'https://server.arcgisonline.com/ArcGIS/rest/services/World_Imagery/MapServer/tile/{z}/{y}/{x}',
    attribution:
      'Tiles &copy; Esri &mdash; Source: Esri, DigitalGlobe, GeoEye, Earthstar Geographics, CNES/Airbus DS, USDA, USGS, AeroGRID, IGN, and the GIS User Community',
    thumbnail: 'satellite.jpg',
  },
] as BaseLayer[];

// Just changes default layer - using this to simulate a toggle
const toggledLayers = [
  {
    name: 'Satellite',
    url: 'https://server.arcgisonline.com/ArcGIS/rest/services/World_Imagery/MapServer/tile/{z}/{y}/{x}',
    attribution:
      'Tiles &copy; Esri &mdash; Source: Esri, DigitalGlobe, GeoEye, Earthstar Geographics, CNES/Airbus DS, USDA, USGS, AeroGRID, IGN, and the GIS User Community',
    thumbnail: 'satellite.jpg',
  },
  {
    name: 'Map',
    url: 'https://{s}.tile.openstreetmap.org/{z}/{x}/{y}.png',
    attribution: '&copy; <a href="http://osm.org/copyright">OpenStreetMap</a> contributors',
    thumbnail: 'streets.jpg',
  },
] as BaseLayer[];

const component = mount(<BasemapToggle baseLayers={baseMaps} onToggle={toggle} />);

it('renders correctly - defaults on street layer', () => {
  const tree = renderer.create(<BasemapToggle baseLayers={baseMaps} onToggle={toggle} />).toJSON();
  expect(tree).toMatchSnapshot();
});

it('toggle handler called correctly', () => {
  component.prop('onToggle')();
  expect(toggle).toBeCalledTimes(1);
});

it('thumbnail shows satellite layer when on street layer', () => {
  expect(component.find('img').prop('src')).toEqual('satellite.jpg');
});

it('thumbnail shows street layer when on satellite layer', () => {
  const component = mount(<BasemapToggle baseLayers={toggledLayers} onToggle={toggle} />);
  expect(component.find('img').prop('src')).toEqual('streets.jpg');
});

// it('handles updating state correctly', ()=> {
//     component.setState({updating: true});
//     expect(component.find("div").first().prop("className")).toEqual("basemap-container view-busy");
// });
