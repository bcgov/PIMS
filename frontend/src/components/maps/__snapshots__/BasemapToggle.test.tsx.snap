// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly - defaults on street layer 1`] = `
<div
  className="basemap-container"
>
  <button
    className="basemap-item current"
  />
  <label
    className="caption"
  >
    Satellite
  </label>
  <div
    className="basemap-item basemap-item-button secondary"
    onClick={[Function]}
  >
    <img
      alt="Map Thumbnail"
      role="presentation"
      src="satellite.jpg"
    />
  </div>
</div>
`;
